[generacmobilelink| Initial Contribution

digitaldan opened the pull request:
This is the initial contribution of the Generac MobileLink binding.  This allows the  Generac line of generators using their MobileLink cloud service to be monitored as things in openHAB.

mhilbush commented:
Very cool! I finally might have a reason to get MobileLink on my standby generator.

digitaldan commented:
Happy to hear someone else might find this useful then!

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/README.md:
```diff
@@ -0,0 +1,88 @@
+| currentAlarmDescription | String      | Current alarm description                 |
+| runHours                | Number:Time | Run hours                                 |
+| exerciseHours           | Number:Time | Exercise hours                            |
+| fuelType                | Switch      | Fuel Type                                 |
```
good catch, it should be a decimal.  I'm not actually sure what it means, i reversed the API, so i can only guess at the meaning

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/README.md:
```diff
@@ -0,0 +1,88 @@
+| runHours                | Number:Time | Run hours                                 |
+| exerciseHours           | Number:Time | Exercise hours                            |
+| fuelType                | Switch      | Fuel Type                                 |
+| fuelLevel               | Switch      | Fuel Level                                |
```
and i copied and pasted that from the above issue :-)

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>GeneracMobileLink Binding</name>
+	<description>This is the binding for GeneracMobileLink.</description>
```
Agreed.  FYI this was auto generated from our skeleton script

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>GeneracMobileLink Binding</name>
+	<description>This is the binding for GeneracMobileLink.</description>
+	<author>Dan Cunningham</author>
```
Sounds good.  FYI this was auto generated from our skeleton script

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,109 @@
+	<channel-type id="connected">
+		<item-type>Switch</item-type>
+		<label>Connected</label>
+		<state readOnly="true" pattern="%s"/>
```
none i need to remove it.

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/api/GeneratorStatus.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @author Dan Cunningham - Initial contribution
+ */
+public class GeneratorStatus {
```
ahh, i did not know that, and missed your comment until just now.  Thank you for the pointer, that is very helpful.

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/handler/GeneracMobileLinkAccountHandler.java:
```diff
@@ -0,0 +1,244 @@
+
+    private void login() {
+        try {
+            GeneracMobileLinkAccountConfiguration config = getConfigAs(GeneracMobileLinkAccountConfiguration.class);
```
Login is only called once after the binding is initialized, so should not be an issue.

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/discovery/GeneracMobileLinkDiscoveryService.java:
```diff
@@ -0,0 +1,104 @@
+        for (DiscoveryListener discoveryListener : discoveryListeners) {
+            try {
+                discoveryListener.thingDiscovered(this, result);
+            } catch (Exception e) {
```
Not sure why had that there, maybe debugging?  its not needed so i have removed it.

digitaldan commented:
Thanks @fwolter for reviewing !   I have finished addressing your comments.

fwolter reviewed (approved):
LGTM

digitaldan commented:
Hi @fwolter sorry for the last min commit.  I had a cheap hack to get around the API returning a 401 without the correct headers that caused the synchronous use of jetty (which most of us use) to throw a generic execution exception.  I was basically parsing the exception message for  specific string since i did not have a response code at that point.  This worked, but was bothering me :-) 

I implemented this again using the asynchronous version along with a compatible future which allows me to properly catch this error.

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/handler/GeneracMobileLinkGeneratorHandler.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Override
+    public void initialize() {
+    }
```
You should at the very least set the thing status to UNKNOWN as part of `initialize()`.

digitaldan replied:
yep

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/factory/GeneracMobileLinkHandlerFactory.java:
```diff
@@ -0,0 +1,95 @@
+public class GeneracMobileLinkHandlerFactory extends BaseThingHandlerFactory {
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_ACCOUNT,
+            THING_TYPE_GENERATOR);
+    private final Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
```
Access to this isn't completely synchronized, so it would be safe to just make the map thread safe.
```suggestion
    private final Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new ConcurrentHashMap<>();
```

digitaldan replied:
yep

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,109 @@
+		<label>Fuel Level</label>
+		<state readOnly="true"/>
+	</channel-type>
+	<channel-type id="batteryVoltage">
+		<item-type>String</item-type>
+		<label>Battery Voltage Status</label>
+		<state readOnly="true"/>
+	</channel-type>
```
Is the voltage returned as a number? Or is it a status like "Charging"?

digitaldan replied:
It's a String, like "good".  Unfortunately some of the variations on the possible return values are unknown to me.  There is no public documentation on this API, and i could find no other projects on the internet who have tried an integration, so this was me truly reverse engineering the API.

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/discovery/GeneracMobileLinkDiscoveryService.java:
```diff
@@ -0,0 +1,95 @@
+ * @author Dan Cunningham - Initial contribution
+ */
+@NonNullByDefault
+public class GeneracMobileLinkDiscoveryService implements DiscoveryService {
```
That would have been better, yes :-)

digitaldan commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,109 @@
+		<state readOnly="true"/>
+	</channel-type>
+	<channel-type id="statusDate">
+		<item-type>String</item-type>
+		<label>Last Status Date</label>
+		<state readOnly="true"/>
```
You caught me :-) I'll give it a try, as mentioned in another comment, i had to reverse the API, i'm not sure how the time string might change given geographic regions or timezones, but its probably the right thing to do to at least try.

digitaldan replied:
actually now that i look at this closely, its seems the API just returns a format like "12/19/2020" , so i'm not sure if a datetime object really makes sense.  I'm not sure what time zone this is referring to.

digitaldan replied:
I changed this to a datetime starting the the beginning of the local day.

digitaldan commented:
Thanks for the feedback @cpmeister !  I have addressed your comments in the latest commit.

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/handler/GeneracMobileLinkAccountHandler.java:
```diff
@@ -0,0 +1,267 @@
+                    handleErrorResponse(result);
+                }
+            }
+        } catch (InterruptedException e) {
+        }
```
You should silently consume InterruptedExceptions. If this thread is interrupted it usually means that the system is requesting that it stops run as soon as it is able. To prevent the callers of your method from assuming that execution can continue as normal you should have your method throw the InterruptedException instead.

Let the InterruptedException travel as high up the call stack as possible to guarantee an expedient thread completion.

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/java/org/openhab/binding/generacmobilelink/internal/handler/GeneracMobileLinkAccountHandler.java:
```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    private void generatorDiscovered(GeneratorStatusDTO generator) {
+        DiscoveryResult result = DiscoveryResultBuilder
+                .create(new ThingUID(GeneracMobileLinkBindingConstants.THING_TYPE_GENERATOR, getThing().getUID(),
+                        String.valueOf(generator.gensetID)))
+                .withLabel("MobileLink Generator " + generator.generatorName)
+                .withProperty("generatorId", String.valueOf(generator.gensetID))
+                .withRepresentationProperty("generatorId").withBridge(getThing().getUID()).build();
+        discoveryService.thingDiscovered(result);
+    }
```
This logic should probably be moved into the GeneracMobileLinkDiscoveryService instead.

cpmeister commented on bundles/org.openhab.binding.generacmobilelink/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,109 @@
+			<bridge-type-ref id="account"/>
+		</supported-bridge-type-refs>
+		<label>MobileLink Generator</label>
+		<description>Generator</description>
```
Maybe make the description longer than the label... just saying.

cpmeister reviewed (changes requested):
From what little I know about generators, I know that they must be run periodically as part of their maintenance. Is it possible through the mobile link to turn on/off the generator remotely so that such maintenance is scheduled through openHAB?

digitaldan commented:
I have addressed your comments in my latest commit, thanks. 

> From what little I know about generators, I know that they must be run periodically as part of their maintenance. Is it possible through the mobile link to turn on/off the generator remotely so that such maintenance is scheduled through openHAB?

The API allows for read only access, their mobile client is read only as well.

digitaldan commented:
Updated for 3.1.0-snapshot
